compilers/concrete-optimizer/concrete-optimizer/src/dag/unparametrized.rs

```diff
@@ -353,74 +339,342 @@ impl OperationDag {
         }
     }
 
-    fn infer_out_precision(&self, op: &UnparameterizedOperator) -> Precision {
+    fn infer_out_precision(&self, op: &Operator) -> Precision {
         match op {
             Operator::Input { out_precision, .. }
             | Operator::Lut { out_precision, .. }
             | Operator::UnsafeCast { out_precision, .. }
             | Operator::Round { out_precision, .. } => *out_precision,
             Operator::Dot { inputs, .. } | Operator::LevelledOp { inputs, .. } => {
-                self.out_precisions[inputs[0].i]
+                self.dag.out_precisions[inputs[0].0]
             }
         }
     }
 }
 
+/// A type containing a Directed Acyclic Graph of operators.
+///
+/// This is the major datatype used to encode a module in the optimizer. It is equivalent to an fhe
+/// module in the frontend, or to a mlir module in the compiler: it can contain multiple separated
+/// circuits which are optimized together.
+///
+/// To add a new circuit to the dag, one should create a [`DagBuilder`] associated to the circuit
+/// using the [`Dag::builder`] method, and use the `DagBuilder::add_*` methods.
+///
+/// # Note:
+///
+/// For ease of use in tests, it is also possible to add operators on an anonymous circuit (`_`)
+/// directly on a [`Dag`] object itself, using the `Dag::add_*` methods.
+#[derive(Clone, PartialEq, Debug)]
+#[must_use]
+pub struct Dag {
+    pub(crate) operators: Vec<Operator>,
+    // Collect all operators output shape
+    pub(crate) out_shapes: Vec<Shape>,
+    // Collect all operators output precision
+    pub(crate) out_precisions: Vec<Precision>,
+    // Collect whether operators are tagged as outputs
+    pub(crate) output_tags: Vec<bool>,
+    // Collect the circuit the operators are associated with
+    pub(crate) circuit_tags: Vec<String>,
+}
+
+impl Default for Dag {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Dag {
+    /// Instantiate a new dag.
+    pub fn new() -> Self {
+        Self {
+            operators: vec![],
+            out_shapes: vec![],
+            out_precisions: vec![],
+            output_tags: vec![],
+            circuit_tags: vec![],
+        }
+    }
+
+    /// Returns a builder for the circuit named `circuit`.
+    pub fn builder<A: AsRef<str>>(&mut self, circuit: A) -> DagBuilder {
+        DagBuilder {
+            dag: self,
+            circuit: circuit.as_ref().into(),
+        }
+    }
+
+    pub fn add_input(
```

Comment:
Is those methods are still used? Are we using DEFAULT_CIRCUIT?

Reply:
I mean the building methods, for me only the builder ones should be used.

Reply:
Those methods are not exposed in `concrete-optimizer-cpp`, I just left them to avoid changing all the tests of the optimizer.